File: example/flux_example_raw/src/image_pub.cpp
#include "flux/ros/publisher.hpp"

#include <rclcpp/rclcpp.hpp>

#include <chrono>
#include <cstddef>
#include <cstdint>
#include <memory>

// The path a node takes with no .msg: no adapter, no schema, no header fields. The frame is a
// bare uint8 volume and the publisher names its dtype and shape itself. Kept next to the three
// adapter nodes so the cost of having no schema is visible -- this one carries no stamp and no
// frame_id, because there is nowhere to put them.
class ImageRawPublisher : public rclcpp::Node
{
public:
  ImageRawPublisher()
  : Node("flux_raw_image_pub"), pub_(*this, kTopic, flux::kNoSchema, kSlotSize, kSlotCount)
  {
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(1.0 / kRateHz)),
      [this]() { tick(); });
  }

private:
  static constexpr const char * kTopic = "image_raw";
  static constexpr std::uint64_t kWidth = 1920;
  static constexpr std::uint64_t kHeight = 1200;
  static constexpr std::uint64_t kChannels = 3;
  static constexpr double kRateHz = 30.0;
  static constexpr std::uint32_t kSlotCount = 8;
  static constexpr std::size_t kDataBytes = static_cast<std::size_t>(kWidth * kHeight * kChannels);
  static constexpr std::uint32_t kSlotSize = static_cast<std::uint32_t>(kDataBytes);

  void tick()
  {
    flux::WriteSlot w = pub_.loan(flux::DType::U8, {kHeight, kWidth, kChannels});
    if (!w.valid()) {
      return;
    }

    auto * px = static_cast<std::uint8_t *>(w.data());
    for (std::size_t i = 0; i < kDataBytes; ++i) {
      px[i] = static_cast<std::uint8_t>(i + phase_);
    }

    // Backpressure is a dropped frame on a best-effort transport, so it is not an error.
    // Anything else does not clear on its own.
    if (const flux::Published p = w.commit(); flux::faulted(p)) {
      RCLCPP_ERROR_ONCE(get_logger(), "commit refused: %s", flux::to_string(p));
    }
    ++phase_;
  }

  flux::ros::Publisher pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::uint8_t phase_ = 0;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImageRawPublisher>());
  rclcpp::shutdown();
  return 0;
}

File: example/flux_example_raw/flux_example_raw/image_pub.py
#!/usr/bin/env python3
"""Publish a bare uint8 volume with no .msg and no adapter.

ros2 run flux_example_raw image_pub.py
"""

import flux.ros
import numpy as np
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node


class ImageRawPublisher(Node):

    TOPIC = "image_raw"
    WIDTH = 1920
    HEIGHT = 1200
    CHANNELS = 3
    DATA_BYTES = WIDTH * HEIGHT * CHANNELS
    SLOT_COUNT = 8
    RATE_HZ = 30.0

    def __init__(self):
        super().__init__("flux_raw_image_pub")
        self.ramp = np.arange(self.DATA_BYTES, dtype=np.uint8).reshape(
            self.HEIGHT, self.WIDTH, self.CHANNELS
        )
        self.phase = 0
        self.publisher = flux.ros.Publisher(
            self,
            self.TOPIC,
            fingerprint=flux.NO_SCHEMA,
            slot_size=self.DATA_BYTES,
            slot_count=self.SLOT_COUNT,
        )
        self.timer = self.create_timer(1.0 / self.RATE_HZ, self.tick)

    def tick(self):
        # The array already states dtype and shape, so nothing restates them here. This is what
        # the C++ raw node spells out by hand in loan().
        loan = self.publisher.loan(self.ramp.shape, dtype="uint8")
        if loan:
            np.add(self.ramp, np.uint8(self.phase), out=loan.array)
            self.report(loan.commit())
        self.phase = (self.phase + 1) % 256

    def report(self, published):
        """Backpressure is a dropped frame on a best-effort transport, so it is not an error.
        Anything else does not clear on its own."""
        if flux.faulted(published):
            self.get_logger().error(f"publish refused: {published}")


def main(args=None):
    rclpy.init(args=args)
    node = ImageRawPublisher()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
